File: 1255-maximum-score-words-formed-by-letters/1255-maximum-score-words-formed-by-letters.java
class Solution {
    int solve(String[] words, int[] count, int[] score, int idx) {
        int max = 0;
        for(int i = idx; i<words.length; i++) {
            int ans = 0;
            boolean flag = true;
            for(char c : words[i].toCharArray()) {
                count[c - 'a']--;
                ans += score[c - 'a'];
                if(count[c - 'a'] < 0) flag = false;
            }
            if(flag) {
                ans += solve(words, count, score, i + 1);
                max = Math.max(ans, max);
            }
            for(char c : words[i].toCharArray()) {
                count[c - 'a']++;
                ans = 0;
            }
        }
        return max;
    }

    public int maxScoreWords(String[] words, char[] letters, int[] score) {
        int n = words.length, m = letters.length, p = score.length;
        if(words == null || n == 0 || letters == null || m == 0 || score == null || p == 0) return 0;
        int[] count = new int[p];
        for(char c : letters) count[c - 'a']++;

        int ans = solve(words, count, score, 0);
        return ans;
    }
}

File: 1255-maximum-score-words-formed-by-letters/1255-maximum-score-words-formed-by-letters.js
/**
 * @param {string[]} words
 * @param {character[]} letters
 * @param {number[]} score
 * @return {number}
 */

function solve(words, count, score, idx) {
    let ans = 0;
    for (let i = idx; i < words.length; i++) {
        let curr = 0;
        let flag = true;
        for (let c of words[i]) {
            count[c.charCodeAt(0) - 'a'.charCodeAt(0)]--;
            curr += score[c.charCodeAt(0) - 'a'.charCodeAt(0)];
            if (count[c.charCodeAt(0) - 'a'.charCodeAt(0)] < 0) flag = false;
        }
        if (flag) {
            curr += solve(words, count, score, i + 1);
            ans = Math.max(ans, curr);
        }
        for (let c of words[i]) {
            count[c.charCodeAt(0) - 'a'.charCodeAt(0)]++;
        }
        curr = 0;
    }
    return ans;
}

var maxScoreWords = function(words, letters, score) {
    let n = words.length, m = letters.length, p = score.length;
    if (n === 0 || m === 0 || p === 0) return 0;
    let count = new Array(p).fill(0);
    for (let c of letters) {
        count[c.charCodeAt(0) - 'a'.charCodeAt(0)]++;
    }
    return solve(words, count, score, 0);
};